test/audit_logs/test_audit_log.py

```diff
@@ -0,0 +1,94 @@
+import datetime
+import pytest
+import six
+from girder import auditLogger
+from girder.models.file import File
+from girder.models.folder import Folder
+from girder.models.upload import Upload
+
+
+@pytest.fixture
+def recordModel():
+    from girder.plugins.audit_logs import Record
+    yield Record()
+
+
+@pytest.fixture
+def resetLog():
+    yield auditLogger
```

Comment:
Should we be mocking the actual logger object instead? Rather than our tests taking a `resetLog` they would take a `mockAuditLog` or something similar?

Reply:
Depends what we want to assert. I was more focusing on asserting that our handler was working, your strategy would be for asserting that the log messages are being dispatched.

Reply:
In principle, this test could be written as a unit-ish tests with multiple independent parts: one which:
  1. Mocked the audit logger
  2. Made some HTTP requests
  3. Asserted that entries were added to the mocked log
and another which:
  1. Mocked the audit logger
  2. Directly added some entries to the mocked log
  3. Asserted that the plugin handled these entries as expected

Note that without mocking, these two parts would not be parallelizable with Pytest, as both parts interacting with the common global logger could cause failures.

Of course, as the current test does not examine the log data structure directly, is more of an end-to-end test, not a unit test (which is fine). However, although the non-mocking fixture is fine under the current structure, I think it's important that (wherever possible) we use fixtures to mock, rather than mutate-as-a-side-effect. Firstly, it will make this fixture usable as-is if we ever want to do more direct unit testing of this code. Secondofly, people are very likely going to copy-paste these sort of fixures as examples when writing additional tests, so we should prevent potential anti-patterns from spreading.

Reply:
> Note that without mocking, these two parts would not be parallelizable with Pytest, as both parts interacting with the common global logger could cause failures.

In my experience, when pytest is run in parallel, it does not run multiple tests in parallel in the same python process. It forks off several subprocesses and schedules the tests onto them, but they are run serially within each process. Is there some multithreaded pytest execution mode I'm not aware of?

Reply:
> Is there some multithreaded pytest execution mode I'm not aware of?

That would be available after installing the concurrent plugin:

https://github.com/reverbc/pytest-concurrent